File: src/workspace_workbench/core/review.py
from __future__ import annotations

from typing import Any, Mapping

from .errors import WorkbenchError
from .git import GitClient


def compare_entry(git: GitClient, repository: Mapping[str, Any], workspace_id: str, target_ref: str) -> dict[str, Any]:
    entry: dict[str, Any] = {
        "workspaceId": workspace_id, "repoPath": repository["repoPath"],
        "branch": "", "headShort": "", "targetBranch": target_ref,
        "targetHeadShort": "", "relation": "unknown", "dirty": False,
        "unpushed": False, "commitCount": None, "paths": [], "issues": [],
        "changes": {"files": 0, "additions": 0, "deletions": 0, "binaryFiles": 0},
    }
    try:
        head = git.head()
        entry["headShort"] = (head or "")[:8]
        entry["branch"] = git.branch() or ""
        entry["dirty"] = bool(git.status_porcelain())
        if not head or not target_ref:
            raise WorkbenchError("comparison target is unavailable", code="base_missing")
        target = git.resolve_commit(target_ref)
        entry["targetHeadShort"] = target[:8]
        behind, ahead = map(int, git.run(["rev-list", "--left-right", "--count", f"{target}...{head}"]).stdout.split())
        entry["relation"] = "already-contained" if ahead == 0 else "fast-forward-candidate" if behind == 0 else "diverged"
        entry["commitCount"] = ahead
        _, upstream = git.upstream()
        entry["unpushed"] = bool(upstream and int(git.run(["rev-list", "--count", f"{upstream}..{head}"]).stdout))
        files = git.changed_files("branch", base=target)
        entry["paths"] = [item.path for item in files]
        entry["changes"] = {
            "files": len(files), "additions": sum(item.additions or 0 for item in files),
            "deletions": sum(item.deletions or 0 for item in files),
            "binaryFiles": sum(item.binary for item in files),
        }
    except WorkbenchError as error:
        entry["issues"].append(error.as_dict())
    return entry


def aggregate(entries_by_repository: Mapping[str, list[dict[str, Any]]]) -> list[dict[str, Any]]:
    result = []
    for repo_path, entries in entries_by_repository.items():
        owners: dict[str, list[str]] = {}
        for entry in entries:
            for path in entry.pop("paths", []):
                owners.setdefault(path, []).append(entry["workspaceId"])
        overlaps = [{"path": path, "workspaceIds": ids} for path, ids in owners.items() if len(set(ids)) > 1]
        attention = bool(overlaps or any(item["issues"] or item["dirty"] or item["relation"] in {"unknown", "diverged"} for item in entries))
        status = "needs-review" if attention else "already-contained" if all(item["relation"] == "already-contained" for item in entries) else "fast-forward-candidate"
        result.append({
            "repoPath": repo_path, "entries": entries, "overlaps": overlaps,
            "status": status, "requiresReview": attention,
            "aggregate": {
                "commits": sum(item["commitCount"] or 0 for item in entries),
                **{key: sum(item["changes"][key] for item in entries) for key in ("files", "additions", "deletions")},
            },
        })
    return result


def brief(repositories: list[dict[str, Any]]) -> dict[str, Any]:
    by_repository = {}
    for repository in repositories:
        lines = [f"Repository: {repository['repoPath']}", f"Status: {repository['status']}"]
        for entry in repository["entries"]:
            lines.append(f"- {entry['workspaceId']}: {entry['branch']} → {entry['targetBranch']} ({entry['relation']})")
            lines.extend(f"  Issue: {issue['code']}" for issue in entry["issues"])
        lines.extend(f"Overlap: {item['path']}" for item in repository["overlaps"])
        by_repository[repository["repoPath"]] = "\n".join(lines)
    return {"text": "\n\n".join(by_repository.values()), "byRepository": by_repository}

File: paseo-plugin/server/backend/review.ts
import { Workspaces } from "./workspaces.ts";
import { Git } from "./git.ts";
import { count, protocol } from "./observation.ts";
import { issue, now, WorkbenchError, type Json } from "./storage.ts";
export async function compare(
  workspaces: Workspaces,
  params: Json,
  includeBrief = false,
) {
  if (!Array.isArray(params.workspaceIds) || !params.workspaceIds.length)
    throw new WorkbenchError("workspace_required", "workspaceIds are required");
  const targets = params.targetRefs || {};
  if (typeof targets !== "object" || Array.isArray(targets))
    throw new WorkbenchError("request_invalid", "targetRefs must be an object");
  const ids = [...new Set(params.workspaceIds.map(String))],
    groups: Record<string, Json[]> = Object.create(null);
  for (const id of ids) {
    const workspace = workspaces.get(id);
    if (!workspace.managed)
      throw new WorkbenchError(
        "live_workspace_not_reviewable",
        "live workspace is not reviewable",
      );
    for (const repo of workspace.repositories) {
      const targetRef = String(targets[repo.repoPath] || repo.baseRef || ""),
        git = new Git(repo.worktreePath, workspaces.config.gitTimeout);
      const entry: Json = {
        workspaceId: id,
        repoPath: repo.repoPath,
        branch: "",
        headShort: "",
        targetBranch: targetRef,
        targetHeadShort: "",
        relation: "unknown",
        dirty: false,
        unpushed: false,
        commitCount: null,
        paths: [],
        issues: [],
        changes: count([]),
      };
      try {
        const head = await git.head();
        entry.headShort = (head || "").slice(0, 8);
        entry.branch = (await git.branch()) || "";
        entry.dirty = !!(await git.status(repo.role === "gitlink-root")).length;
        if (!head || !targetRef)
          throw new WorkbenchError(
            "base_missing",
            "comparison target unavailable",
          );
        const target = await git.commit(targetRef),
          [behind, ahead] = (
            await git.text([
              "rev-list",
              "--left-right",
              "--count",
              `${target}...${head}`,
            ])
          )
            .split(/\s+/)
            .map(Number);
        entry.targetHeadShort = target.slice(0, 8);
        entry.relation =
          ahead === 0
            ? "already-contained"
            : behind === 0
              ? "fast-forward-candidate"
              : "diverged";
        entry.commitCount = ahead;
        const [, upstream] = await git.upstream();
        entry.unpushed =
          !!upstream &&
          Number(
            await git.text(["rev-list", "--count", `${upstream}..${head}`]),
          ) > 0;
        const files = await git.files("branch", target);
        entry.paths = files.map((file) => file.path);
        entry.changes = count(files);
      } catch (error) {
        entry.issues.push(issue(error));
      }
      (groups[repo.repoPath] ||= []).push(entry);
    }
  }
  const repositories = Object.entries(groups).map(([repoPath, entries]) => {
    const owners: Record<string, string[]> = Object.create(null);
    for (const entry of entries) {
      for (const path of entry.paths)
        (owners[path] ||= []).push(entry.workspaceId);
      delete entry.paths;
    }
    const overlaps = Object.entries(owners)
      .filter(([, ids]) => new Set(ids).size > 1)
      .map(([path, workspaceIds]) => ({ path, workspaceIds }));
    const requiresReview =
      !!overlaps.length ||
      entries.some(
        (entry) =>
          entry.issues.length ||
          entry.dirty ||
          ["unknown", "diverged"].includes(entry.relation),
      );
    return {
      repoPath,
      entries,
      overlaps,
      requiresReview,
      status: requiresReview
        ? "needs-review"
        : entries.every((entry) => entry.relation === "already-contained")
          ? "already-contained"
          : "fast-forward-candidate",
      aggregate: {
        commits: entries.reduce((n, e) => n + (e.commitCount || 0), 0),
        ...Object.fromEntries(
          ["files", "additions", "deletions"].map((key) => [
            key,
            entries.reduce((n, e) => n + e.changes[key], 0),
          ]),
        ),
      },
    };
  });
  const issues = repositories.flatMap((repo) =>
    repo.entries.flatMap((entry) => entry.issues),
  );
  const result: Json = {
    schemaVersion: protocol,
    workspaceIds: ids,
    repositories,
    overlaps: repositories.flatMap((repo) =>
      repo.overlaps.map((overlap) => ({ repoPath: repo.repoPath, ...overlap })),
    ),
    testEvidence: null,
    issues,
    observation: {
      state: issues.length ? "partial" : "ready",
      observedAt: now(),
      issues,
    },
  };
  if (includeBrief) {
    const byRepository = Object.fromEntries(
      repositories.map((repo) => [
        repo.repoPath,
        [
          `Repository: ${repo.repoPath}`,
          `Status: ${repo.status}`,
          ...repo.entries.flatMap((entry) => [
            `- ${entry.workspaceId}: ${entry.branch} → ${entry.targetBranch} (${entry.relation})`,
            ...entry.issues.map((e: Json) => `  Issue: ${e.code}`),
          ]),
          ...repo.overlaps.map((item) => `Overlap: ${item.path}`),
        ].join("\n"),
      ]),
    );
    result.brief = {
      text: Object.values(byRepository).join("\n\n"),
      byRepository,
    };
  }
  return result;
}
